File: android/app/src/main/java/com/kgeneye/eye/auth/KGenAuthClient.kt
package com.kgeneye.eye.auth

import com.kgeneye.eye.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Locale

class KGenAuthClient(
    private val http: OkHttpClient = OkHttpClient(),
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val mediaJson = "application/json".toMediaType()
    private val base = BuildConfig.KGEN_AUTH_BASE.ifBlank {
        "https://wvsixcvsfndhoygbkzkj.supabase.co/functions/v1/auth-mobile"
    }
    private val appKey = BuildConfig.KGEN_APP_KEY

    suspend fun login(email: String, password: String): AuthSession =
        decodeSession(call("login", mapOf("email" to normalizeEmail(email), "password" to password)))

    suspend fun signup(
        email: String,
        password: String,
        fullName: String?,
        country: String?,
        city: String?,
        referralCode: String?,
    ): AuthSession = decodeSession(
        call(
            "signup",
            mapOf(
                "email" to normalizeEmail(email),
                "password" to password,
                "full_name" to fullName,
                "country" to country,
                "city" to city,
                "referral_code" to referralCode,
            ),
        ),
    )

    suspend fun google(idToken: String, referralCode: String?): AuthSession =
        decodeSession(call("google", mapOf("id_token" to idToken, "referral_code" to referralCode)))

    suspend fun refresh(refreshToken: String): AuthSession =
        decodeSession(call("refresh", mapOf("refresh_token" to refreshToken)))

    suspend fun forgotPassword(email: String) {
        call("forgot-password", mapOf("email" to normalizeEmail(email)))
    }

    private suspend fun call(action: String, body: Map<String, Any?>, bearer: String? = null): String =
        withContext(Dispatchers.IO) {
            if (appKey.isBlank() || appKey.contains("<")) {
                throw IllegalStateException("Missing KGEN_APP_KEY in local.properties")
            }

            val payload = json.encodeToString(JsonObject.serializer(), body.toJsonObject())
            val requestBuilder = Request.Builder()
                .url("$base?action=$action")
                .post(payload.toRequestBody(mediaJson))
                .header("Content-Type", "application/json")
                .header("x-app-key", appKey)
            bearer?.let { requestBuilder.header("Authorization", "Bearer $it") }

            http.newCall(requestBuilder.build()).execute().use { response ->
                val text = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw IllegalStateException(parseError(text) ?: "Auth error HTTP ${response.code}")
                }
                text
            }
        }

    private fun decodeSession(text: String): AuthSession =
        try {
            json.decodeFromString(AuthSession.serializer(), text)
        } catch (_: SerializationException) {
            throw IllegalStateException("Could not decode auth response")
        }

    private fun parseError(text: String): String? =
        runCatching {
            val obj = json.parseToJsonElement(text) as? JsonObject
            obj?.get("error")?.let { (it as? JsonPrimitive)?.content }
        }.getOrNull()

    private fun normalizeEmail(email: String): String =
        email.trim().lowercase(Locale.US)

    private fun Map<String, Any?>.toJsonObject(): JsonObject = buildJsonObject {
        for ((key, value) in this@toJsonObject) {
            when (value) {
                null -> Unit
                is String -> put(key, JsonPrimitive(value))
                is Number -> put(key, JsonPrimitive(value))
                is Boolean -> put(key, JsonPrimitive(value))
                is JsonElement -> put(key, value)
                else -> put(key, JsonPrimitive(value.toString()))
            }
        }
    }
}

File: EgoCapture/Services/Auth/KGenAuthClient.swift
import Foundation

final class KGenAuthClient {
    static let shared = KGenAuthClient()

    private let baseURL: URL
    private let appKey: String
    private let session: URLSession
    private let decoder = JSONDecoder()

    init(bundle: Bundle = .main, session: URLSession = .shared) {
        let info = bundle.infoDictionary ?? [:]
        let base = (info["KGEN_AUTH_BASE"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let key = (info["KGEN_APP_KEY"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)

        self.baseURL = URL(string: base?.isEmpty == false ? base! : "https://wvsixcvsfndhoygbkzkj.supabase.co/functions/v1/auth-mobile")!
        self.appKey = key ?? ""
        self.session = session
    }

    func login(email: String, password: String) async throws -> AuthSession {
        let data = try await request(action: "login", body: [
            "email": normalizeEmail(email),
            "password": password,
        ])
        return try decodeSession(data)
    }

    func signup(
        email: String,
        password: String,
        fullName: String?,
        country: String?,
        city: String?,
        referralCode: String?
    ) async throws -> AuthSession {
        let data = try await request(action: "signup", body: [
            "email": normalizeEmail(email),
            "password": password,
            "full_name": fullName,
            "country": country,
            "city": city,
            "referral_code": referralCode,
        ])
        return try decodeSession(data)
    }

    func google(idToken: String, accessToken: String?, referralCode: String?) async throws -> AuthSession {
        let data = try await request(action: "google", body: [
            "id_token": idToken,
            "access_token": accessToken,
            "referral_code": referralCode,
        ])
        return try decodeSession(data)
    }

    func refresh(refreshToken: String) async throws -> AuthSession {
        let data = try await request(action: "refresh", body: ["refresh_token": refreshToken])
        return try decodeSession(data)
    }

    func me(accessToken: String) async throws -> AuthUser {
        let data = try await request(action: "me", body: [:], bearer: accessToken)
        struct Envelope: Codable { let user: AuthUser }
        do {
            return try decoder.decode(Envelope.self, from: data).user
        } catch {
            throw AuthError.decoding
        }
    }

    func forgotPassword(email: String) async throws {
        _ = try await request(action: "forgot-password", body: ["email": normalizeEmail(email)])
    }

    private func request(action: String, body: [String: Any?], bearer: String? = nil) async throws -> Data {
        guard !appKey.isEmpty, !appKey.contains("<") else {
            throw AuthError.missingConfiguration("KGEN_APP_KEY")
        }

        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "action", value: action)]
        guard let url = components.url else { throw AuthError.invalidResponse }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(appKey, forHTTPHeaderField: "x-app-key")
        if let bearer {
            request.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONSerialization.data(
            withJSONObject: body.compactMapValues { $0 },
            options: []
        )

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            #if DEBUG
            throw AuthError.server("Auth \(action) transport failed: \(error.localizedDescription) url=\(url.absoluteString)")
            #else
            throw AuthError.server("Auth request failed: \(error.localizedDescription)")
            #endif
        }
        guard let http = response as? HTTPURLResponse else { throw AuthError.invalidResponse }

        if http.statusCode == 401 && action == "login" {
            throw AuthError.invalidCredentials
        }
        guard (200...299).contains(http.statusCode) else {
            let rawBody = String(data: data, encoding: .utf8) ?? "<non-utf8>"
            #if DEBUG
            throw AuthError.server("Auth \(action) failed (HTTP \(http.statusCode)): \(rawBody)")
            #else
            throw AuthError.server(errorMessage(from: data) ?? "Auth error HTTP \(http.statusCode)")
            #endif
        }
        return data
    }

    private func decodeSession(_ data: Data) throws -> AuthSession {
        do {
            return try decoder.decode(AuthSession.self, from: data)
        } catch {
            throw AuthError.decoding
        }
    }

    private func errorMessage(from data: Data) -> String? {
        guard
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let message = json["error"] as? String
        else { return nil }
        return message
    }

    private func normalizeEmail(_ email: String) -> String {
        email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }
}
